clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/lifecycle/InstanceTerminationLifecycleAgent.java

```diff
@@ -0,0 +1,220 @@
+          .withWaitTimeSeconds(properties.getWaitTimeSeconds())
+      );
+
+      receiveMessageResult.getMessages().forEach(message -> {
+        try {
+          LifecycleMessage lifecycleMessage = objectMapper.readValue(message.getBody(), LifecycleMessage.class);
+
+          if (SUPPORTED_LIFECYCLE_TRANSITION.equalsIgnoreCase(lifecycleMessage.lifecycleTransition)) {
+            Task originalTask = TaskRepository.threadLocalTask.get();
+            try {
+              TaskRepository.threadLocalTask.set(
+                Optional.ofNullable(originalTask).orElse(new DefaultTask(InstanceTerminationLifecycleAgent.class.getSimpleName()))
+              );
+              handleMessage(lifecycleMessage, TaskRepository.threadLocalTask.get());
+            } finally {
+              TaskRepository.threadLocalTask.set(originalTask);
+            }
+          }
+        } catch (IOException e) {
+          log.error("Unable to convert NotificationMessage (body: {})", message.getBody(), e);
+        }
+
+        deleteMessage(amazonSQS, queueId, message);
+        messagesProcessed.incrementAndGet();
+      });
+
+      if (receiveMessageResult.getMessages().isEmpty()) {
+        // No messages
+        break;
+      }
+    }
+
+    log.info("Processed {} messages (queueARN: {})", messagesProcessed.get(), queueARN.arn);
+  }
+
+  private void handleMessage(LifecycleMessage message, Task task) {
+    List<String> instanceIds = Collections.singletonList(message.ec2InstanceId);
+
+    EnableDisableInstanceDiscoveryDescription description = new EnableDisableInstanceDiscoveryDescription();
+    description.setCredentials(getAccountCredentialsById(message.accountId));
+    description.setRegion(queueARN.region);
+    description.setAsgName(message.autoScalingGroupName);
+    description.setInstanceIds(instanceIds);
+
+    discoverySupport.get().updateDiscoveryStatusForInstances(
+      description, task, "handleLifecycleMessage", DiscoveryStatus.Disable, instanceIds
+    );
+  }
+
+  private static void deleteMessage(AmazonSQS amazonSQS, String queueUrl, Message message) {
+    try {
+      amazonSQS.deleteMessage(queueUrl, message.getReceiptHandle());
+    } catch (ReceiptHandleIsInvalidException e) {
+      log.warn("Error deleting lifecycle message, reason: {} (receiptHandle: {})", e.getMessage(), message.getReceiptHandle());
+    }
+  }
+
+  private NetflixAmazonCredentials getAccountCredentialsById(String accountId) {
+    return (NetflixAmazonCredentials) accountCredentialsProvider.getAll()
+      .stream()
+      .filter(c -> c.getAccountId().equals(accountId))
+      .findFirst()
+      .orElseThrow((Supplier<RuntimeException>) () -> {
+        return new RuntimeException(String.format("Unable to find AmazonCredentials by id (id: %s)", accountId));
+      });
+  }
+
+  private static String ensureQueueExists(AmazonSQS amazonSQS, ARN queueARN, String sourceARN, Set<String> allAccountIds) {
+    String queueUrl;
+    try {
```

Comment:
I believe the createQueue(name) will return the queue URL if it already exists. http://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_CreateQueue.html.